### ISSUE TITLE
Start Backend Test Dev

### ISSUE THREAD

**Issue by mtreacy002** [anitab-org/bridge-in-tech-backend#19]
## Description
As a student,
I need create and populate data models classes,
so that I can move on to the next tasks specified in the First Coding Phase Week 1.

## Mocks
N/A

### Update [Required]
- [ ] must have enough pip dependencies (aka requirements.txt) at least to start the app 
- [ ] must have remote server established as well as local server
- [ ] able to run project and see swagger UI 
- [ ] must have .env.template
- [ ] must have migration scripts for initial migration
- [ ] must have .gitignore, .pylintrc, and .travis.yml
- [ ] must have functions if user opted to create and populate UserModel, MentorshipRelationModel, TasksListModel, TaskCommentModel, UserExtensionModel, PersonalBackgroundModel, OrganizationModel, ProgramModel, MentorshipRelationExtensionModel
- [ ] must have basic tests setup e.g. test_config.py and other simple tests

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by all mentors.

## Estimation
2 days

**Comment by mtreacy002** [comment]
Hi @anitab-org/bridgeintech-maintainers. I've created models for the BridgeInTech DB which consists of 2 schemas, bit schema and public (aka ms schema).
Here's the screenshot of the DB structure on pgadmin4

<img width="320" alt="Screen Shot 2020-05-19 at 10 28 15 pm" src="https://user-images.githubusercontent.com/29667122/82328850-bc2fbf00-9a23-11ea-8ef7-26b0c9ba6b98.png">

And here's the Swagger UI running on local server from running `python run.py` command (note that I haven't started any API routing here).

<img width="770" alt="Screen Shot 2020-05-19 at 10 28 36 pm" src="https://user-images.githubusercontent.com/29667122/82328975-f1d4a800-9a23-11ea-8c6a-2b2797a0c8a4.png">

Here's the SQL statements for each tables:

**MS Schema (public):**

* users

<img width="480" alt="Screen Shot 2020-05-19 at 10 58 15 pm" src="https://user-images.githubusercontent.com/29667122/82329257-5ee83d80-9a24-11ea-92bc-445f5ec0f885.png">

* mentorship_relations

<img width="554" alt="Screen Shot 2020-05-19 at 11 00 07 pm" src="https://user-images.githubusercontent.com/29667122/82329334-863f0a80-9a24-11ea-9970-21e4466b8db9.png">

* tasks_list

<img width="461" alt="Screen Shot 2020-05-19 at 11 01 03 pm" src="https://user-images.githubusercontent.com/29667122/82329449-acfd4100-9a24-11ea-848d-69e218f2bfc6.png">


* tasks_comments

<img width="556" alt="Screen Shot 2020-05-19 at 11 01 52 pm" src="https://user-images.githubusercontent.com/29667122/82329503-c1d9d480-9a24-11ea-827f-3190807cb219.png">

**bitschema**

* users_extension

<img width="523" alt="Screen Shot 2020-05-19 at 11 03 50 pm" src="https://user-images.githubusercontent.com/29667122/82329674-082f3380-9a25-11ea-94f1-62171cefc9e2.png">

* personal_backgrounds

<img width="559" alt="Screen Shot 2020-05-19 at 11 04 56 pm" src="https://user-images.githubusercontent.com/29667122/82329794-2dbc3d00-9a25-11ea-92ca-53d9fdbd462c.png">

* organizations

<img width="528" alt="Screen Shot 2020-05-19 at 11 05 38 pm" src="https://user-images.githubusercontent.com/29667122/82329851-47f61b00-9a25-11ea-937a-336bc5d410b7.png">

* programs

<img width="557" alt="Screen Shot 2020-05-20 at 9 46 30 am" src="https://user-images.githubusercontent.com/29667122/82389215-df3b8c80-9a7e-11ea-9164-7a23d93a6ace.png">

* mentorship_relation_extension_model

<img width="668" alt="Screen Shot 2020-05-19 at 11 08 54 pm" src="https://user-images.githubusercontent.com/29667122/82330109-bdfa8200-9a25-11ea-9862-fc9f11cac926.png">

Note: the `_model` at the end of this table just appear out of nowhere. In the code base, I use the exact same style as other tables as table name. Weird... 🤔. Maybe someone can lend me another pair of 👀 and spot why this happens?

<img width="741" alt="Screen Shot 2020-05-19 at 11 11 35 pm" src="https://user-images.githubusercontent.com/29667122/82330452-34977f80-9a26-11ea-9454-aafd57ca7efd.png">


I will commit and push the codes to my remote forked repo. But as this is only for a test, I probably won't open a PR until we finalise it 😉 . The branch is called [test-schema](https://github.com/mtreacy002/bridge-in-tech-backend/tree/test-schema).

Look forward to your feedback 😆

**Comment by mtreacy002** [comment]
PS: tomorrow I'm gonna work on migration script using Alembic and perhaps if time permits, will populate the tables with mock data. @ramitsawhney27, can I populate the data directly on pgadmin4 by running sql statements so not by placing the query inside the code (inside the code will get populated every time we're running the app)? This is coz I'm afraid the later will cause issue coz the data already exist.

Or, should I do it with test cases for db models? Coz this way the mock data will only get inserted while the test is running and will be removed at the end of each test.

Can you please advise?

**Comment by ramitsawhney27** [comment]
Dummy data insertion via pgadmin4 is good. We don't need tests as discussed in the weekly meeting.

**Comment by mtreacy002** [comment]
Notes to self:

The following changes need to be made to address `datetime` inconsistencies from mentorship-backend code base (discussed in mentorship-backend [PR#556](https://github.com/anitab-org/mentorship-backend/pull/556) and [issue#608](https://github.com/anitab-org/mentorship-backend/issues/608):

* **app/database/models/user.py**
  * **register**
    * current: time.time()
       recorded in: UTC+0
    * required: datetime.utcnow()
      recorded in: UTC+0 
<img width="565" alt="Screen Shot 2020-05-23 at 1 11 15 pm" src="https://user-images.githubusercontent.com/29667122/82720654-e79efc00-9cf8-11ea-8755-e29387ef480e.png">

     direct impact: 
     * **app/schedulers/delete_unverified_users_cron_job.py**

<img width="592" alt="Screen Shot 2020-05-23 at 1 16 08 pm" src="https://user-images.githubusercontent.com/29667122/82720535-d6a1bb00-9cf7-11ea-9e50-17c7de6329af.png">

     * **app/tests/cron_job_functions/test_delete_unverified_users_cron.py**

<img width="672" alt="Screen Shot 2020-05-23 at 1 21 02 pm" src="https://user-images.githubusercontent.com/29667122/82720604-4d3eb880-9cf8-11ea-9951-a5808ac09f03.png">



* **app/api/dao/user.py**
  * **confirm_register**
    * current: datetime.now()
      recorded in: local UTC (action user's UTC offset)
    * required: datetime.utcnow()
      to be recorded in: UTC+0

<img width="586" alt="Screen Shot 2020-05-23 at 1 00 41 pm" src="https://user-images.githubusercontent.com/29667122/82720281-714cca80-9cf5-11ea-9e76-bacb661b0596.png">

    
More to be added ....

**Comment by mtreacy002** [comment]
Update:
Hi @anitab-org/bridgeintech-maintainers. I've populated the MS+BIT tables with mock data using the code I just pushed to [test-schema](https://github.com/mtreacy002/bridge-in-tech-backend/tree/test-schema) branch on my forked repo.  I set it so the application will automatically populate the database when you run it. 
You can test it by setting up your own postgresql and virtual environment locally (just like we set up development environment for Mentorship System). After you install dependencies (`pip install -r requirements.txt`) and run the app with `python run.py` command, the database will be auto-populated with dummy data. you can view them either with pgadmin4 or running query inside psql shell. Atm I'm working on a db_query_mock.py to auto-run few queries when running the application. Will let you know when I'm done.

Let me know if there's anything unclear on the setting up 😉

**Comment by mtreacy002** [comment]
Update:
Hi @anitab-org/bridgeintech-maintainers.
So, I've tried the following tests to see if BIT could send requests to MS API service.
I've placed the query in BIT codes ([db_query_mock](https://github.com/mtreacy002/bridge-in-tech-backend/blob/test-schema/app/database/db_query_mock.py)) to automatically send request POST /register when running application (as I haven't got BIT swagger UI set up properly for API endpoints)

I've done the tests with 3 different conditions:

* **Test 1. Running both service locally (2 APIs + 2 DBs):**
1. run MS API on local port 4000 with database sqlite local_data.db
1. run BIT API on local port 5000 with database postgresql bit_schema local database

* **Test 2. MS API on Heroku, BIT on local port 5000 (2 APIs and 2 DBs)**
1. connect to MS API on Heroku server (which means it has whatever database connected to Heroku as set in dev server - this would be AnitaB's elephantsql posgres MS, I pressume)
2. run BIT API as per above (running locally on port 5000 with postgresql bit_schema local database 

* **Test 3. Running both APIs locally with one target DB as what we want to achieve with this project (as last discussed on [issue#10](https://github.com/anitab-org/bridge-in-tech-backend/issues/10#issuecomment-631705821)).**
1. run MS API on local port 4000 with database target postgresql bit_schema
1. run BIT API on local port 5000 with database the same target database postgresql bit_schema

The results are as follows:

* **Test 1 result (local MS API + local BIT API + 2 different DBs)**
1. Successfully registered and new user data is recorded in MS sqlite local_data.db (not recorded in BIT postgresql bit_schema db)
<img width="1680" alt="Screen Shot 2020-05-25 at 2 50 53 pm" src="https://user-images.githubusercontent.com/29667122/82781271-b19a7d00-9e9c-11ea-914b-1e66d6acb8fd.png">

* **Test 2 result (heroku MS API + local BIT API + 2 different DBs)**
1. Successfully registered and new user data is recorded in MS heroku db (not recorded in BIT postgresql bit_schema db)

as I don't have access to debugger console when connecting to heroku server, I took the below screenshot instead.
Right screenshot is successful login on MS heroku API whereas Left screenshot is successful login on MS local API (sqlite3 local_data.db).

<img width="1680" alt="Screen Shot 2020-05-25 at 2 55 15 pm" src="https://user-images.githubusercontent.com/29667122/82781537-646adb00-9e9d-11ea-98ff-f1fa5a60858c.png">

* **Test 3 result (MS local API + BIT local API with one target DB, postgresql bit_schema db.**

<img width="1680" alt="Screen Shot 2020-05-25 at 2 59 27 pm" src="https://user-images.githubusercontent.com/29667122/82781884-55385d00-9e9e-11ea-820e-660a0d3e5e92.png">

This test 3 raised sqlalchemy error which I think caused by the non-existence BIT schemas inside the current code base of develop branch on MS (mentorship-backend repo). So regardless that the postgresql it's targeting as database already have BIT schemas in place.

@ramitsawhney27. Can you please advise how we should solve this issue? Thanks beforehand.

**Comment by ramitsawhney27** [comment]
You could discuss this with Isabel to see how this can be included in mentorship system.

**Comment by mtreacy002** [comment]
@isabelcosta, can you share your thoughts?

**Comment by mtreacy002** [comment]
Update:
Hi @anitab-org/bridgeintech-maintainers. I've just created and pushed the alembic script to support flask-migration to the [test-schema branch](https://github.com/mtreacy002/bridge-in-tech-backend/tree/test-schema). 

<img width="803" alt="Screen Shot 2020-05-26 at 11 33 31 am" src="https://user-images.githubusercontent.com/29667122/82852238-bf580d00-9f45-11ea-843a-e46191aa7b3c.png">

We can also use it to do migration on mentorship-backend if we've decided to go ahead with incorporating bitschema in mentorship-backend code base.
Note: Since we have existing data in current db, extra care need to be taken so that the migration won't affect them (if we choose to keep these data). Not sure how (export import to csv?? 🤷 )
